src/smockit/smockit.ts

```diff
@@ -195,55 +216,113 @@ export const smockit = async (
-        ? mockFn.will.returnValue(...params)
+      mockFn.will?.returnValue instanceof Function
+        ? await mockFn.will.returnValue(...params)
         : mockFn.will.returnValue
 
     let encodedReturnValue: string = '0x'
     if (rawReturnValue !== undefined) {
-      try {
-        encodedReturnValue = this.interface.encodeFunctionResult(fn, [
-          rawReturnValue,
-        ])
-      } catch (err) {
-        if (err.code === 'INVALID_ARGUMENT') {
+      if (mockFn.will?.resolve === 'revert') {
+        if (typeof rawReturnValue !== 'string') {
+          throw new Error(
+            `Smock: Tried to revert with a non-string (or non-bytes) type: ${typeof rawReturnValue}`
+          )
+        }
+
+        if (rawReturnValue.startsWith('0x')) {
+          encodedReturnValue = rawReturnValue
+        } else {
+          const errorface = new ethers.utils.Interface([
+            {
+              inputs: [
+                {
+                  name: '_reason',
+                  type: 'string',
+                },
+              ],
+              name: 'Error',
+              outputs: [],
+              stateMutability: 'nonpayable',
+              type: 'function',
+            },
+          ])
+
+          encodedReturnValue = errorface.encodeFunctionData('Error', [
+            rawReturnValue,
+          ])
+        }
+      } else {
+        if (fn === null) {
+          encodedReturnValue = rawReturnValue
+        } else {
           try {
-            encodedReturnValue = this.interface.encodeFunctionResult(
-              fn,
-              rawReturnValue
-            )
-          } catch {
-            if (typeof rawReturnValue !== 'string') {
-              throw new Error(
-                `Could not properly encode mock return value for ${fn.name}`
-              )
-            }
+            encodedReturnValue = this.interface.encodeFunctionResult(fn, [
+              rawReturnValue,
+            ])
+          } catch (err) {
+            if (err.code === 'INVALID_ARGUMENT') {
+              try {
+                encodedReturnValue = this.interface.encodeFunctionResult(
+                  fn,
+                  rawReturnValue
+                )
+              } catch {
+                if (typeof rawReturnValue !== 'string') {
+                  throw new Error(
+                    `Could not properly encode mock return value for ${fn.name}`
+                  )
+                }
 
-            encodedReturnValue = rawReturnValue
+                encodedReturnValue = rawReturnValue
+              }
+            } else {
+              throw err
+            }
           }
-        } else {
-          throw err
         }
       }
+    } else {
+      if (fn === null) {
+        encodedReturnValue = '0x'
+      } else {
+        encodedReturnValue = '0x' + '00'.repeat(2048)
+      }
     }
 
     return {
-      resolve: mockFn.will.resolve,
+      resolve: mockFn.will?.resolve,
+      functionName: fn ? fn.name : null,
```

Comment:
`functionName` and `rawReturnValue` currently unused. Will be using them to improve how we do call assertions.